### PR TITLE
Add support for extra date and bpm tags in music files

### PR DESCRIPTION
**Pull request by the-black-eagle**
## Description
This PR gives Kodi the ability to read the original release date frames from music files, allowing albums to be sorted by the original release date instead of the album release date (as v18 generally does).  Many users have mis-tagged their music files on purpose in order to obtain the correct original release date when sorting by years but this will no longer be necessary now.  Albums (and songs) can be sorted by either the original release date or the year of the album release.  
This also adds a new setting to the music library where users can choose to use the original release date at all times rather than album release date.  When enabled, all dates used for sorting and display will only use the original release year.  Whether the setting is enabled or not, both dates are available via the new list & player items.  All dates are localized to the short date format selected in Kodi's settings.
The PR also adds reading the BPM tag and methods for sorting and displaying it.

## Motivation and Context
Without deliberately mis-tagging your music files, it has been impossible to sort and display albums in chronological order.  I have many albums that are actually re-releases of original albums that I would like to see listed in the correct release date order but I also like to keep my tagging correct and until now, the two have not gone hand in hand. 
There has been a discussion about Kodi having the ability to do this on the forum https://forum.kodi.tv/showthread.php?tid=313952 which was first raised over three years ago.
BPM handling was added because I have seen several requests on the forum for it and it seemed appropriate to add it to the tag reader at the same time I was adding the extra date handling.

## How Has This Been Tested?
This has been tested on Ubuntu 18.04 with both local sqlite and shared MySQL databases.  Dates added to the database were checked against the tags in the files.  All types of tagging were tested (id3v2.3/2.4, ape, flac, mp4 etc) and all relevant documentation checked to ensure compatibility across tagging types.

## Screenshots (if appropriate):
Before (Highlighted album was originally released in 1978)
<a href="https://ibb.co/54v6CLg">screenshot135</a>

With new sort by original date (albums in correct chronological order)
<a href="https://ibb.co/YXCfBWf">screenshot136</a>

Album info dialog with new setting OFF showing the album (re-)release year
<a href="https://ibb.co/pPCdK7t">screenshot137</a>

And with the setting enabled, showing the original release year
<a href="https://ibb.co/2FV3D1w">screenshot138</a>
## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [X] All new and existing tests passed

### CODE REVIEW

File: xbmc/music/Album.cpp [hunk 1]
```diff
@@ -484,11 +487,21 @@ bool CAlbum::Load(const TiXmlElement *album, bool append, bool prioritise)
...
-  XMLUtils::GetInt(album,"year",iYear);
+  XMLUtils::GetString(album, "releasedate", strReleaseDate);
+  StringUtils::Trim(strReleaseDate);  // @todo: validate ISO8601 format
+  // Support old style <year></year> for backwards compatibility
+  if (strReleaseDate.empty())
+  {
+    int year;
+    if (XMLUtils::GetInt(album, "year", year));
+    if (year > 0)
+      strReleaseDate = StringUtils::Format("%04i", year);
```
Comment (DaveTBlake) [inline xbmc/music/Album.cpp]: Misplaced `;` don't really need if at all
This is the compile warning issue on some platforms hence jenkins fail

Comment (the-black-eagle) [reply xbmc/music/Album.cpp]: Sorted.  Thanks.

**Comment by jjd-uk** [comment]
I'm curious on the SmartPlaylist part of this. I see mention of BPM here https://github.com/xbmc/xbmc/pull/17437/commits/8fb8a640e076aa6fdd0c554892630d97bb067d07#diff-5107b0bbec8dc3030c39d1f567612fe7R125

Does this mean you can create a SmartPlaylist with a rule such as **BPM greater than 120**

If that is possible, then my next question would is there any Node support so you could create a **High Tempo** node for example using a BPM greater than rule.

**Comment by the-black-eagle** [comment]
Yes, you could use that as a rule.  There are no "built in" nodes but you can certainly create one yourself.

Smart playlist editor
<a href="https://ibb.co/w4YjbgW">screenshot139</a>

Result ( I don't have a lot of stuff tagged with bpm, just for testing purposes really)
<a href="https://ibb.co/PjBnbnB">screenshot140</a>

You can take the smartplaylist rule and convert it into a node eg
```
<?xml version='1.0' encoding='UTF-8'?>
<node type="filter">
    <label>High Tempo</label>
    <content>songs</content>
    <match>all</match>
    <rule field="bpm" operator="greaterthan">
        <value>129</value>
    </rule>
    <order direction="ascending">bpm</order>
</node>
```
<a href="https://ibb.co/3Nk8TDN">screenshot141</a>

**Comment by jjd-uk** [comment]
Yep that's all I was asking is that whether a custom node made by the user would also be possible using a BPM rule (assuming that it could be done in SmartPlaylist). So thanks for confirming all is possible, great work :)

**Comment by DaveTBlake** [comment]
Test build available http://mirrors.kodi.tv/test-builds/windows/win64/KodiSetup-20200306-1b101377-PR17437-merge-x64.exe

**Review by DaveTBlake — APPROVED** [review]
This looks good @the-black-eagle , certainly robust enough in testing to be in the nightlies so going to hit merge :)

**Pull request merged**